src/Common/src/TypeSystem/IL/Stubs/PInvokeILEmitter.cs

```diff
@@ -175,8 +162,63 @@ public static MethodIL EmitIL(MethodDesc method, PInvokeILEmitterConfiguration p
                 return emitter.Link(method);
             }
         }
+
+        private bool IsStubRequired()
+        {
+            MethodDesc method = _methodData.TargetMethod;
+            Debug.Assert(method.IsPInvoke);
+
+            if (MarshalHelpers.UseLazyResolution(method, _methodData.ImportMetadata.Module, _methodData.PInvokeILEmitterConfiguration))
+            {
+                return true;
+            }
+            if ((_methodData.ImportMetadata.Attributes & PInvokeAttributes.SetLastError) == PInvokeAttributes.SetLastError)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < _marshallers.Length; i++)
+            {
+                if (_marshallers[i].isMarshallingRequired())
+                    return true;
+            }
+            return false;
+        }
+
+        public static bool IsStubRequired(MethodDesc method, PInvokeILEmitterConfiguration pinvokeILEmitterConfiguration)
+        {
+            return new PInvokeILEmitter(method, pinvokeILEmitterConfiguration).IsStubRequired();
```

Comment:
Yes, I agree.  The same non-trivial work is again done during EmitIL. It would be nice to reuse this object. How about adding a cache from MethodDesc -> PInvokeILEmitter  so that it can be reused?